src/ultk/effcomm/rate_distortion.py

```diff
@@ -0,0 +1,191 @@
+
+        prior: array of shape `|meanings|` representing the communicative need distribution
+
+    Returns:
+        a tuple of floats `(complexity, accuracy, distortion)`, s.t.
+
+            `complexity`: the complexity of the lexicon I[M:W], in bits
+
+            `accuracy`: the accuracy of the lexicon I[W:U], in bits
+
+            `distortion`: the distortion E[DKL[ M || M_hat ]] = I[M:U] - I[W:U], in bits
+    """
+    args = [language, prior, meaning_dists]
+    return ib_encoder_to_point(
+        *args[1:], language_to_ib_encoder_decoder(*args).values()
+    )
+
+
+def language_to_ib_encoder_decoder(
+    language: Language,
+    prior: np.ndarray,
+    meaning_dists: np.ndarray,
+) -> dict[str, np.ndarray]:
+    """Convert a Language, a mapping of words to meanings, to IB encoder, q(w|m) and IB decoder q(m|w).
+
+    A Bayesian decoder chooses an interpretation according to p(meaning | word), where
+
+    $P(m | w) = \\frac{P(M | W) \cdot P(M)} { P(W) }$
+
+    Furthermore, we will require that each word w is deterministically interpreted as meaning $\hat{m}$ as follows:
+
+    $\hat{m}_{w}(u) = \sum_m p(m|w) \cdot m(u)$
+
+    See https://github.com/nogazs/ib-color-naming/blob/master/src/ib_naming_model.py#L40.
+
+    Args:
+        language: the lexicon from which to infer a speaker (encoder).
+
+        prior: communicative need distribution
+
+    Returns:
+        a dict of the form
+        {
+            "encoder": np.ndarray of shape `(|meanings|, |words|)`,
+            "decoder": np.ndarray of shape `(|words|, |meanings|)`,
+        }
+    """
+    # In the IB framework, the encoder _can_ be a literal speaker and the decoder is a bayes optimal listener.
+    speaker = LiteralSpeaker(language)
+    speaker.weights = rows_zero_to_uniform(speaker.normalized_weights())
+    listener = Listener(language)
+    listener.weights = ib_optimal_decoder(speaker.weights, prior, meaning_dists)
+    return {
+        "encoder": speaker,
+        "decoder": listener,
+    }
+
+
+def ib_encoder_to_point(
+    meaning_dists: np.ndarray,
+    prior: np.ndarray,
+    encoder: np.ndarray,
+    decoder: np.ndarray = None,
+) -> tuple[float]:
+    """Return (complexity, accuracy, comm_cost) IB coordinates.
+
+    Args:
+        meaning_dists: array of shape `(|meanings|, |meanings|)` representing the distribution over world states given meanings.
+
+        prior: array of shape `|meanings|` representing the communicative need distribution
+
+        encoder: array of shape `(|meanings|, |words|)` representing P(W | M)
+
+        decoder: array of shape `(|words|, |meanings|)` representing P(M | W).  By default is None, and the Bayesian optimal decoder will be inferred.
+    """
+    if decoder is None:
+        decoder = ib_optimal_decoder(encoder, prior, meaning_dists)
+
+    complexity = information_cond(prior, encoder)
+    accuracy = MI(meaning_dists @ joint(encoder, prior))
+    distortion = MI(joint(meaning_dists, prior)) - accuracy
```

Comment:
a comment or two on these lines, or a bit more detail in the docstring, would be helpful